From 89632ecc8a8d406e00593359019c4fa622cbf184 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Thu, 17 Jan 2013 18:15:57 -0800
Subject: [PATCH 328/337] netdev/octeon-ethernet: Silence rmmod warnings on
 CN68XX

Based On SDK 3.0.0-482

We need to disable SSO_ERR[FPE] while removing memory from SSO.

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Corey Minyard <cminyard@mvista.com>
Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 drivers/net/ethernet/octeon/ethernet-rx.c |   13 +++++++++++--
 1 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/octeon/ethernet-rx.c b/drivers/net/ethernet/octeon/ethernet-rx.c
index 6d1b840..aa74fb8 100644
--- a/drivers/net/ethernet/octeon/ethernet-rx.c
+++ b/drivers/net/ethernet/octeon/ethernet-rx.c
@@ -44,6 +44,7 @@
 #endif /* CONFIG_XFRM */
 
 #include <asm/octeon/octeon.h>
+#include <asm/octeon/octeon-hw-status.h>
 
 #include "ethernet-defines.h"
 #include "octeon-ethernet.h"
@@ -510,12 +511,16 @@ void cvm_oct_rx_shutdown1(void)
 	union cvmx_sso_fpage_cnt fpage_cnt;
 	int num_to_transfer, count, i;
 	void *mem;
-
-
+	const int sso_fpe_bit = 45;
 
 	if (!OCTEON_IS_MODEL(OCTEON_CN68XX))
 		return;
 
+	/* Spurious FPE errors will happen doing this cleanup.
+	 * Disable the indication.
+	 */
+	octeon_hw_status_disable(CVMX_SSO_ERR, 1ull << sso_fpe_bit);
+
 	sso_cfg.u64 = cvmx_read_csr(CVMX_SSO_CFG);
 	sso_cfg.s.rwen = 0;
 	sso_cfg.s.rwq_byp_dis = 1;
@@ -581,4 +586,8 @@ void cvm_oct_rx_shutdown1(void)
 	cvmx_write_csr(CVMX_SSO_CFG, sso_cfg.u64);
 	kmem_cache_destroy(cvm_oct_kmem_sso);
 	cvm_oct_kmem_sso = NULL;
+
+	/* Clear any FPE indicators, and reenable. */
+	cvmx_write_csr(CVMX_SSO_ERR, 1ull << sso_fpe_bit);
+	octeon_hw_status_enable(CVMX_SSO_ERR, 1ull << sso_fpe_bit);
 }
-- 
1.7.5.4

